src/Microsoft.AspNetCore.Blazor.Browser/Interop/JavaScriptInvoke.cs

```diff
@@ -0,0 +1,253 @@
+﻿using System;
+using System.Linq;
+using System.Reflection;
+using System.Text;
+using System.Threading.Tasks;
+
+namespace Microsoft.AspNetCore.Blazor.Browser.Interop
+{
+    /// <summary>
+    /// TODO
+    /// </summary>
+    public class JavaScriptInvoke
+    {
+        /// <summary>
+        /// TODO
+        /// </summary>
+        /// <param name="methodOptions"></param>
+        /// <param name="methodArguments"></param>
+        /// <returns></returns>
+        public static string InvokeDotnetMethod(string methodOptions, string methodArguments)
+        {
+            Console.WriteLine(methodOptions);
+            Console.WriteLine(methodArguments);
+            var options = JsonUtil.Deserialize<MethodOptions>(methodOptions);
+            var argumentDeserializer = GetOrCreateArgumentDeserializer(options);
+            var invoker = GetOrCreateInvoker(options, argumentDeserializer);
+
+            var result = invoker(methodArguments);
+            if (options.Async != null && !(result is Task))
+            {
+                throw new InvalidOperationException($"'{options.Method.Name}' in '{options.Type.TypeName}' must return a Task.");
+            }
+            if (result is Task taskResult)
```

Comment:
What's the plan for if `options.Async == null` but `result is Task`? (i.e., someone is trying to call an async method but using the sync API)

Should we throw in this case? Might help developers to discover what they are doing wrong.

Reply:
Yes.